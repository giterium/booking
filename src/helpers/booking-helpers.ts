import moment from "moment";
import styles from "../css/booking.module.css";
import { store } from '../configureStore';
import {TypeBooking} from "../reducers/booking-reducers";
//import {RootState} from "../reducers";

export function isGoodDiapazon (startDate, endDate, id_room, curIdBooking = '') {

    const booking:TypeBooking[] = store.getState().booking;
    for(const curBooking of booking) {
        if(
            (
                moment(timenull(startDate)).isAfter(timenull(curBooking.startDate)) && moment(timenull(startDate)).isBefore(timenull(curBooking.endDate))
                ||
                moment(timenull(endDate)).isAfter(timenull(curBooking.startDate)) && moment(timenull(endDate)).isBefore(timenull(curBooking.endDate))
                ||
                moment(timenull(startDate)).isSameOrBefore(timenull(curBooking.startDate)) && moment(timenull(endDate)).isSameOrAfter(timenull(curBooking.endDate))
                ||
                moment(timenull(startDate)).isSame(timenull(curBooking.startDate)) && moment(timenull(endDate)).isSame(timenull(curBooking.endDate))
            )
            &&
            id_room == curBooking.room
            &&
            curIdBooking != curBooking._id
        ) {
            return false;
        }
    }
    return true;
}

export function timenull(date) {
    if(moment.isMoment(date))
        return date.clone().milliseconds(0).second(0).minutes(0).hours(0);
    else
        return moment(date).clone().milliseconds(0).second(0).minutes(0).hours(0).toDate();
}

export const isBooking = (day, room) => {

    const booking:TypeBooking[] = store.getState().booking;
    let is_booking = false;
    let is_first_booking = false;
    let is_last_booking = false;

    booking.map(curBooking => {
        if (
            moment(timenull(curBooking.startDate)).diff(timenull(day.date)) <= 0
            &&
            moment(timenull(curBooking.endDate)).diff(timenull(day.date)) >= 0
            &&
            curBooking.room == room._id
        ) {
            if(moment(timenull(curBooking.startDate)).diff(timenull(day.date)) == 0) {
                is_first_booking = true;
            }
            if(moment(timenull(curBooking.endDate)).diff(timenull(day.date)) == 0) {
                is_last_booking = true;
            }
            is_booking = true;
        }
    })
    if(is_booking) {
        if(is_last_booking && is_first_booking)
            return styles.isBooking + ' '+styles.isFirstBooking+ ' '+styles.isLastBooking;
        else if(is_first_booking)
            return styles.isBooking + ' '+styles.isFirstBooking;
        else if(is_last_booking)
            return styles.isBooking + ' '+styles.isLastBooking;
        else
            return styles.isBooking
    }
    else
        return '';
}

const commonStylesCell = (day, room) => {
    return isBooking(day, room) + ' d' + day.date.format('DD-MM-YY') + ' ' +styles.dayCell;
}

export const isSelected = (day, room, selected) => {
    if(selected.start.day && !selected.end.day ) {
        if(selected.start.day.format('DD.MM.YY') == day.date.format('DD.MM.YY') && selected.start.room == room._id) {
            return commonStylesCell(day, room)  + ' ' + styles.selected +' '+ styles.firstSelected;
        }
        else {
            return commonStylesCell(day, room) ;
        }
    }
    else if(selected.start.day && selected.end.day && selected.start.room == room._id && selected.end.room == room._id) {
        if(selected.start.day.diff(day.date) <= 0 && selected.end.day.diff(day.date) >= 0) {
            if(selected.start.day.format('DD.MM.YY') == day.date.format('DD.MM.YY')) {
                return commonStylesCell(day, room)  + ' ' + styles.selected + ' ' + styles.firstSelected;
            }
            else if(moment(timenull(selected.end.day)).diff(timenull(day.date)) == 0) {
                return commonStylesCell(day, room) + ' ' + styles.selected + ' ' + styles.lastSelected;
            }
            else
                return commonStylesCell(day, room)  + ' ' + styles.selected;
        }
        else {
            return commonStylesCell(day, room);
        }
    }
    else {
        return commonStylesCell(day, room);
    }
}

