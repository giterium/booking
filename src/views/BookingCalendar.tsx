import React, {useEffect, useState, useContext}  from 'react';
import {TypeRoom} from "../reducers/rooms-reducers";
import {shallowEqual, useSelector} from "react-redux";
import moment, {Moment} from 'moment';
import {WindowContext, IContextProps} from './Booking';
import styles from '../css/booking.module.css';
import {TypeBooking} from "../reducers/booking-reducers";
import {RootState} from "../reducers";
import {isGoodDiapazon, timenull, isBooking, isSelected} from "../helpers/booking-helpers";

export interface TypeDay {
    date: Date;
    display: Date;
}

interface TypeItemSelected {
    day: null | Moment;
    room: string | boolean;
}

interface TypeSelected {
    start: TypeItemSelected;
    end: TypeItemSelected;
}

type CalendarProps = {
    onChangeSelected: (selected: TypeSelected) => void;
    onClickBooking: (id: string) => void;
    selected?: TypeSelected;
    rooms: TypeRoom[];
    booking: TypeBooking[];
};

export const Calendar = (props : CalendarProps) => {
    //const rooms: TypeRoom[] = useSelector((state: RootState) => state.rooms, shallowEqual);
    //const booking: TypeBooking[] = useSelector((state: RootState) => state.booking, shallowEqual);
    const [startDate, setStartDate] = useState<Moment>(moment().milliseconds(0).second(0).minutes(0).hours(0));
    const [endDate, setEndDate] = useState<Moment>(moment().add(13,'days').milliseconds(0).second(0).minutes(0).hours(0));
    const [daysList, setDaysList] = useState<TypeDay[]>([]);
    const [selected, setSelected] = useState<TypeSelected>({
        start:{day:null, room:false},
        end:{day:null, room:false}
    });

    const { openWindow } = useContext<IContextProps>(WindowContext);


    const unSelected = () => {
        setSelected({
            start:{day:false, room:false},
            end:{day:false, room:false}
        });
    }

    const getBookingId = (day, room) => {
        for(const curBooking of props.booking) {
            if (
                curBooking.room == room._id
                &&
                moment(timenull(curBooking.startDate)).diff(timenull(day.date)) == 0
            ) {
                return curBooking._id;
            }
        }
        return 0;
    }

    const getBookingFio = (day, room) => {
        for(const curBooking of props.booking) {
            if (
                curBooking.room == room._id
                &&
                moment(timenull(curBooking.startDate)).diff(timenull(day.date)) == 0
            ) {
                return curBooking.fio;
            }
        }
        return '';
    }

    const getWidthFio = (day, room) => {
        let duration = 0;
        for(const curBooking of props.booking) {
            if (
                curBooking.room == room._id
                &&
                moment(timenull(curBooking.startDate)).diff(timenull(day.date)) == 0
            ) {
                duration = moment(timenull(curBooking.endDate)).diff(moment(timenull(curBooking.startDate)), 'days');
                break;
            }
        }
        if(duration == 1)
            return 50 + 'px';
        else
            return 100*duration-50+'px';
    }

    const clickDayRoom = (day, room) => {
        if(isBooking(day, room) != styles.isBooking && isBooking(day, room) != styles.isBooking + ' '+styles.isFirstBooking+ ' '+styles.isLastBooking) {
            if (!selected.start.day) {
                setSelected({
                    start: {day: day.date, room: room._id},
                    end: {day: false, room: false}
                })
            } else {
                if (!selected.end.day) {
                    if(isGoodDiapazon(selected.start.day, day.date, room._id)) {
                        if (selected.start.day.format('DD.MM.YY') == day.date.format('DD.MM.YY')) {
                            unSelected();
                        } else if (selected.start.room != room._id)
                            unSelected();
                        else {
                            if (day.date < selected.start.day) {
                                setSelected({
                                    start: {day: day.date, room: room._id},
                                    end: selected.start
                                })
                            } else {
                                setSelected({
                                    start: selected.start,
                                    end: {day: day.date, room: room._id}
                                })
                            }

                            /*
                            setOpenWindow(true);
                            setCurrentBooking({
                                _id: 'create',
                                cost: Math.abs(day.date.diff(selected.start.day, "days") * room.cost),
                                fio: '',
                                room: room._id,
                                startDate: selected.start.day.format('YYYY-MM-DD'),
                                endDate: moment(day.date).format('YYYY-MM-DD')
                            });

                             */
                        }
                    }
                    else {
                        unSelected();
                    }
                }
                else {
                    unSelected();
                }
            }
        }
        else {
            unSelected();
        }
    };

    const editBooking = (id) => {

        //const editBooking = props.booking.filter(item => item._id == id)[0];
        props.onClickBooking(id);
        //setCurrentBooking(editBooking);
        //setOpenWindow(true);
    }

    const NextWeek = () => {
        setStartDate(startDate.clone().add(7, 'days'));
        setEndDate(endDate.clone().add(7, 'days'));
    };

    const PrevWeek = () => {
        setStartDate(startDate.clone().subtract(7, 'days'));
        setEndDate(endDate.clone().subtract(7, 'days'));
    };

    useEffect(() => {
        if(selected.end.day)
            props.onChangeSelected(selected);
    }, [selected]);

    useEffect(() => {
        setSelected(props.selected);
    }, [props.selected]);

    useEffect(() => {
        const list:TypeDay[] = [];
        const current = startDate.clone();
        while(current <= endDate){
            list.push({
                date:current.clone(),
                display:current.format('DD.MM')
            });
            current.add(1, 'days');
        }
        setDaysList((prev: TypeDay[]) => prev = list)
    }, [startDate, setEndDate]);

    return <div >

        <div className={styles.shahmatkaBox}>
            <div className={styles.headerRow}>
                <div className={styles.headerCell}></div>
                {daysList.map((day:TypeDay) =>
                    <div key={day.date+'-head'} className={styles.headerCell}>
                        {day.display}
                    </div>
                )}
            </div>
            {props.rooms.map((room:TypeRoom) =>
                <div className={styles.roomRow} key={room.num+'-col'}>
                    <div className={styles.roomCell}>â„– {room.num}</div>
                    {daysList.map((day:TypeDay) =>
                        <div
                            key={room.num+'-'+day.date+'-cell'}
                            onClick={()=>clickDayRoom(day, room)}
                            className={isSelected(day, room, selected)}
                        >
                            <div className={styles.roomSubCell}></div>
                            <div className={styles.roomSubCellLeft}></div>
                            <div
                                onClick={() =>editBooking(getBookingId(day, room))}
                                className={styles.bookingCellFio}
                                style={{width: getWidthFio(day, room), overflow: 'hidden'}}
                            >
                                {getBookingFio(day, room)}
                            </div>
                        </div>
                    )}
                </div>
            )}
            <div className="clear"></div>
        </div>
        <div style={{textAlign: 'center', marginTop: '20px'}}>
            <button onClick={PrevWeek} className={styles.controlButtons}>{'<< Prev'}</button>
            <button onClick={NextWeek} className={styles.controlButtons}>{'Next >>'}</button>
        </div>
        <button onClick={()=>{localStorage.clear(); document.location.reload();}}>{'Clear'}</button>
    </div>
}