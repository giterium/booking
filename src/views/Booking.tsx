import React, { useState, useEffect, useContext } from 'react';
import {useDispatch, shallowEqual, useSelector} from 'react-redux';
import {itemsRoomsFetchData} from '../actions/rooms-actions';
import {clearErrors, deleteBooking, createBooking, itemsBookingFetchData, updateBooking} from '../actions/booking-actions';
import {Calendar} from './BookingCalendar';
import {BookingModal} from './BookingModal';
import {Loading} from '../components/Loading';
import {LoadError} from '../components/LoadError';
import {TitlePage} from '../components/TitlePage';
import {TypeRoom} from '../reducers/rooms-reducers';
import {RootState} from '../reducers';
import {TypeBooking} from '../reducers/booking-reducers';
import moment from 'moment';
import {timenull, updateSelected, setCurrentBooking, changeBooking} from "../utils/booking-utils";

export interface IContextProps {
    openWindow: boolean;
    setOpenWindow: (a: boolean) => void;
    currentBooking: TypeBooking;
}

export const WindowContext = React.createContext({});

export const Booking = () => {
    const bookingIsLoading: boolean = useSelector((state: RootState) => state.bookingIsLoading, shallowEqual);
    const bookingHasErrored: (a: boolean) => boolean = useSelector((state: RootState) => state.bookingHasErrored, shallowEqual);
    const [openWindow, setOpenWindow] = useState(false);
    const currentBooking: TypeBooking = useSelector((state: RootState) => state.currentBooking, shallowEqual);

    const rooms: TypeRoom[] = useSelector((state: RootState) => state.rooms, shallowEqual)
    const booking: TypeBooking[] = useSelector((state: RootState) => state.booking, shallowEqual)
    const dispatch = useDispatch()

    useEffect(() => {
        if(rooms.length == 0)
            dispatch(itemsRoomsFetchData('/api/rooms/getall'))
        if(booking.length == 0)
            dispatch(itemsBookingFetchData('/api/booking/getall'))
    }, []);

    const actionClose = () => {
        dispatch(updateSelected({
            start:{day:null, room:false},
            end:{day:null, room:false}
        }))
    }

    if (bookingHasErrored) {
        dispatch(bookingHasErrored(false))
        return <><LoadError /></>
    }

    if (bookingIsLoading) {
        return <><Loading /></>
    }

    const actionDelete = (id) => {
        if(confirm('Delete?')) {
            dispatch(deleteBooking(
                id,
                booking.findIndex(item => item._id == id)
            ));
        }
    }

    const actionModal = (cost, fio, room, startDate, endDate) => {
        if(currentBooking._id == 'create') {
            dispatch(createBooking({
                _id: '',
                cost: cost,
                fio: fio,
                room: room,
                startDate: startDate,
                endDate: endDate
            }));
        }
        else {
            dispatch(updateBooking({
                _id: currentBooking._id,
                cost: currentBooking.cost,
                fio: fio,
                room: currentBooking.room,
                startDate: currentBooking.startDate,
                endDate: currentBooking.endDate
            }, booking.findIndex(item => item._id == currentBooking._id)));
        }
    }

    const onClickBooking = (id: string) => {
        dispatch(clearErrors())
        const editBooking = booking.filter(item => item._id == id)[0]
        dispatch(setCurrentBooking(editBooking))
        setOpenWindow(true)
    }

    document.title = "Booking"

    return  <div id="wrapper">
                <TitlePage title='Booking' />
                <Calendar
                    onClickBooking={onClickBooking}
                    startDate={timenull(moment())}
                    endDate={timenull(moment().add(13,'days'))}
                />
                <WindowContext.Provider value={{ openWindow, setOpenWindow }}>
                    <BookingModal
                        onChangeBooking={changeBooking}
                        onActionModal={actionModal}
                        onActionDelete={actionDelete}
                        onActionClose={actionClose}
                    />
                </WindowContext.Provider>
            </div>
}