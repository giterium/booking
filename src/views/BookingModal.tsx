import React, {useEffect, useState, useContext}  from 'react';
import {TypeRoom} from "../reducers/rooms-reducers";
import {WindowContext} from './Booking';
import Modal from 'react-modal';
import {TableInput} from "../components/TableInput";
import {TableErrors} from "../components/TableErrors";
import {Button} from "../components/Button";
import {shallowEqual, useSelector, useDispatch} from "react-redux";
import styles from '../css/booking.module.css';
import moment, {Moment} from "moment";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import Select from 'react-select';
import {EventBus} from "../events";
import {RootState} from "../reducers";
import {TypeBooking} from "../reducers/booking-reducers";
import {clearErrors} from "../actions/booking-actions";
import {setCurrentBooking, updateSelected} from "../utils/booking-utils";
import {checkBooking} from "../utils/checkBooking";
import {TypeSelected} from "../reducers/selected-reducers";

const modalStyles = {
    overlay: { zIndex: 100 },
    content : {top : '50%', left : '50%', right : 'auto', bottom : 'auto', marginRight : '-50%', boxShadow : '0 0 10px rgba(0,0,0,0.5)', width : '500px', transform  : 'translate(-50%, -50%)'}
}

const customStyles = {
    control: (styles, isFocused) => ({ ...styles, backgroundColor: 'white', boxShadow: isFocused ? '' : '', border: isFocused ? '0px solid grey' : '0px solid grey' }),
    container: (styles, isFocused) => ({ ...styles, backgroundColor: 'white', border: isFocused ? '1px solid #bfbfbf' : '1px solid #bfbfbf' }),
    singleValue: (styles, isFocused) => ({ ...styles, width: '300px' }),
}

interface TypeRoomsOptions {
    label: number;
    value: string;
}

type ModalProps = {
    //onChangeBooking: (startDate: Moment | Date, endDate: Moment | Date, room: string) => any;
    onActionDelete: (id: string) => void;
    onActionClose?: () => void;
    onActionModal: (cost: string, fio: string, room: string, startDate: Moment | Date, endDate: Moment | Date) => void;
}

export const BookingModal = (props: ModalProps) => {
    const errors: any = useSelector((state: RootState) => state.errors, shallowEqual)
    const rooms: TypeRoom[] = useSelector((state: RootState) => state.rooms, shallowEqual)
    const booking: TypeBooking[] = useSelector((state: RootState) => state.booking, shallowEqual)
    const selected: TypeSelected = useSelector((state: RootState) => state.selected, shallowEqual)
    const currentBooking: TypeBooking = useSelector((state: RootState) => state.currentBooking, shallowEqual)

    const [fio, setFio] = useState('')
    const [startDate, setStartDate] = useState(new Date())
    const [endDate, setEndDate] = useState(new Date())
    const [cost, setCost] = useState('0')
    const [room, setRoom] = useState('')

    const [roomsOptions, setRoomsOptions] = useState([])
    const [defRoomsOptions, setDefRoomsOptions] = useState([])
    const dispatch = useDispatch();
    const { openWindow, setOpenWindow } = useContext(WindowContext)
    EventBus.subscribe('bookingCloseWindow', () => closeWindow())

    const closeWindow = () => {
        if(setOpenWindow)
            setOpenWindow(false)
        if(props.onActionClose)
            props.onActionClose()
    }

    useEffect(() => {
        const roomsOptions: TypeRoomsOptions[] = []
        rooms.map((room) => {
            roomsOptions.push({label: room.num, value: room._id})
        });
        setRoomsOptions(roomsOptions)
    }, [rooms])

    useEffect(() => {
        if(openWindow && currentBooking) {
            setDefRoomsOptions({
                label: rooms.filter(room => room._id == currentBooking.room)[0].num,
                value: currentBooking.room
            })
            setFio(currentBooking.fio)
            setStartDate(currentBooking.startDate)
            setEndDate(currentBooking.endDate)
            setCost(currentBooking.cost)
            setRoom(currentBooking.room)
        }
    }, [openWindow, currentBooking])

    useEffect(() => {
        if (selected.end.room) {
            const curRoom = rooms.filter((item) => item._id == selected.start.room)[0]
            if (typeof curRoom != 'undefined') {
                const cost = Math.abs(moment(selected.start.day).diff(moment(selected.end.day), "days") * parseInt(curRoom.cost))
                dispatch(setCurrentBooking({
                    _id: 'create',
                    cost: cost + '',
                    fio: !openWindow ? '' : fio,
                    room: selected.start.room,
                    startDate: selected.start.day.toDate(),
                    endDate: moment(selected.end.day).toDate()
                }))
                dispatch(updateSelected(selected))
                dispatch(clearErrors())
                setOpenWindow(true)
            }
        }
    }, [selected])

    const changeBooking = (newStartDate, newEndDate, room) => {
        const result = checkBooking(newStartDate, newEndDate, room)
        if(typeof result != 'undefined') {
            if (result === true) {
                const curRoom = rooms.filter((item) => item._id == room)[0];
                setStartDate(newStartDate)
                setEndDate(newEndDate)
                setRoom(room)
                setDefRoomsOptions({
                    label: curRoom.num,
                    value: room
                })
                const cost = Math.abs(moment(startDate).diff(moment(endDate), "days") * parseInt(curRoom.cost))
                setCost(cost+'')
            }
            else {
                const curRoom = rooms.filter((item) => item._id == result[2])[0]
                setStartDate(result[0])
                setEndDate(result[1])
                setRoom(result[2])
                setDefRoomsOptions({
                    label: curRoom.num,
                    value: result[2]
                })
                const cost = Math.abs(moment(startDate).diff(moment(endDate), "days") * parseInt(curRoom.cost))
                setCost(cost+'')
            }
        }
    }

    return <div id="root">
        <Modal
            isOpen={rooms[0] && rooms[0]._id == 'enzymeOpenWindow' ? true : openWindow}
            style={modalStyles}
            ariaHideApp={false}
        >
            <h2 className="titleModal">{(typeof currentBooking == 'undefined' || currentBooking._id == 'create') ? 'Adding a reservation' : currentBooking.fio} </h2>
            <br />
            <table className={styles.tableModal}>
                <tbody>
                    <TableInput title="Name" className="usualInput" name="fio" changeUpdate={(e) => setFio(e.target.value)} defaultValue={fio} />
                    <TableErrors errors={errors && errors.fio ? errors.fio : ''} />

                    <tr><td  className={styles.cellModal}>Date of entry: </td><td>
                        <DatePicker
                            className="startDatePicker"
                            locale="ru"
                            selected={new Date(startDate)}
                            onChange={date => changeBooking(date, endDate, room)}
                            dateFormat="dd.MM.yyyy"
                        />
                    </td></tr>
                    <TableErrors errors={errors && errors.startDate ? errors.startDate : ''} />

                    <tr><td className={styles.cellModal}>Date of departure: </td><td>
                        <DatePicker
                            className="endDatePicker"
                            locale="ru"
                            selected={new Date(endDate)}
                            onChange={date => changeBooking(startDate, date, room)}
                            dateFormat="dd.MM.yyyy"
                        />
                    </td></tr>
                    <TableErrors errors={errors && errors.endDate ? errors.endDate : ''} />

                    <tr><td className={styles.cellModal}>Room: </td><td>
                        <div style={{width: '100px'}}>
                            <Select
                                classNamePrefix='list'
                                className='listRooms'
                                styles={customStyles}
                                value={defRoomsOptions}
                                onChange={(e)=>changeBooking(startDate, endDate, e.value)}
                                name="id_room"
                                options={roomsOptions}
                            />
                        </div>
                    </td></tr>
                    <TableErrors errors={errors && errors.room ? errors.room : ''} />

                    <tr><td className={styles.cellModal}>
                        Cost: </td><td><span className="cost">{cost}</span> $
                    </td></tr>
                </tbody>
            </table>

            <Button onClick={() => props.onActionDelete((currentBooking ? currentBooking._id : ''))} className={
                ((typeof currentBooking == 'undefined' || currentBooking._id == 'create') ? 'hidden' : '')+
                ' deleteButton '+styles.deleteButton
            } title="Delete" />

            <div className={styles.modalBoxButton}>
                <Button
                    className="actionButton"
                    onClick={() => {props.onActionModal(cost, fio, room, startDate, endDate)}}
                    title={(typeof currentBooking == 'undefined' || currentBooking._id == 'create') ? 'Add' : 'Save'}
                />
                <Button onClick={() => closeWindow()} className="closeButton" title="Cancel" />
            </div>
        </Modal>
    </div>
}