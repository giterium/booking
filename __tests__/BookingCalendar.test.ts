import React from "react";
import { Calendar } from "../src/views/BookingCalendar";
import "regenerator-runtime/runtime";
import {timenull} from "../src/helpers/booking-helpers";
import moment from "moment";

describe("Booking Calendar component", () => {
    let rooms = [{
        _id: '233422323',
        num: 23,
        building: 2,
        capacity: 2,
        cost: '100'
    }];

    const mockOnChangeSelected = jest.fn();
    const mockOnClickBooking = jest.fn();
    const props = {
        rooms,
        booking: [{_id: '123', fio: '123'}],
        onChangeSelected: mockOnChangeSelected,
        onClickBooking: mockOnClickBooking,
        startDate: timenull(moment()),
        endDate: timenull(moment().add(13,'days')),
        selected: {
            start:{day:false, room:false},
            end:{day:false, room:false}
        }
    };

    it("should render Booking Calendar component", () => {
        const props = {rooms};
        const wrapper = render(<Calendar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render roomRow", () => {
        const wrapper = mount(<Calendar {...props} />);
        expect(wrapper.find(".roomRow")).toHaveLength(1);
    });

    it("should call onChangeSelected", () => {
        const wrapper = mount(<Calendar {...props} />);
        wrapper.find(".cell").first().simulate('click')
        wrapper.find(".cell").last().simulate('click')
        expect(mockOnChangeSelected.mock.calls.length).toBe(1);
    });

    it("should call onChangeSelected", () => {
        const wrapper = mount(<Calendar {...props} />);
        wrapper.find(".bookingCellFio").first().simulate('click')
        expect(mockOnClickBooking.mock.calls.length).toBe(1);
    });
});