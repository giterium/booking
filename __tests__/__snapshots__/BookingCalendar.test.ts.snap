// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking Calendar component should render Booking Calendar component 1`] = `
"<Provider store={{...}}>
  <Calendar onClickBooking={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} startDate={{...}} endDate={{...}}>
    <div>
      <div className=\\"shahmatkaBox\\">
        <div className=\\"headerRow\\">
          <div className=\\"headerCell\\" />
          <div className=\\"headerCell\\">
            01.01
          </div>
          <div className=\\"headerCell\\">
            02.01
          </div>
          <div className=\\"headerCell\\">
            03.01
          </div>
          <div className=\\"headerCell\\">
            04.01
          </div>
          <div className=\\"headerCell\\">
            05.01
          </div>
          <div className=\\"headerCell\\">
            06.01
          </div>
          <div className=\\"headerCell\\">
            07.01
          </div>
          <div className=\\"headerCell\\">
            08.01
          </div>
          <div className=\\"headerCell\\">
            09.01
          </div>
          <div className=\\"headerCell\\">
            10.01
          </div>
          <div className=\\"headerCell\\">
            11.01
          </div>
          <div className=\\"headerCell\\">
            12.01
          </div>
          <div className=\\"headerCell\\">
            13.01
          </div>
          <div className=\\"headerCell\\">
            14.01
          </div>
        </div>
        <div className=\\"roomRow\\">
          <div className=\\"roomCell\\">
            â„– 
            23
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d01-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d02-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d03-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d04-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d05-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d06-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d07-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d08-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d09-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d10-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d11-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d12-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d13-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
          <div onClick={[Function: onClick]} className=\\" cell d14-01-21 dayCell\\">
            <div className=\\"roomSubCell\\" />
            <div className=\\"roomSubCellLeft\\" />
            <div onClick={[Function: onClick]} className=\\"bookingCellFio\\" style={{...}} />
          </div>
        </div>
        <div className=\\"clear\\" />
      </div>
      <div style={{...}}>
        <Button onClick={[Function: PrevWeek]} className=\\"controlButtons\\" title=\\"<< Prev\\">
          <button onClick={[Function: PrevWeek]} className=\\"controlButtons\\">
            &lt;&lt; Prev
          </button>
        </Button>
        <Button onClick={[Function: NextWeek]} className=\\"controlButtons\\" title=\\"Next >>\\">
          <button onClick={[Function: NextWeek]} className=\\"controlButtons\\">
            Next &gt;&gt;
          </button>
        </Button>
      </div>
      <Button onClick={[Function: onClick]} title=\\"Clear\\">
        <button onClick={[Function: onClick]} className={[undefined]}>
          Clear
        </button>
      </Button>
    </div>
  </Calendar>
</Provider>"
`;
